### ISSUE TITLE
flotilla tasks are stuck as pending in the flotilla backend

### ISSUE THREAD

**Issue by aaj-st** [stitchfix/flotilla-os#95]
## TL;DR
Tasks started via the flotilla UI get started and show up on the ECS console and their logs on cloudwatch.  No output on flotilla UI because the flotilla backend thinks the task is PENDING

<img width="843" alt="48508120-9019f900-e802-11e8-8686-9452dff36a96" src="https://user-images.githubusercontent.com/44782696/48639998-3c3a1c00-e98a-11e8-8c6a-584ac4183c61.png">

## Details

I followed the [instructions](https://github.com/stitchfix/flotilla-os/blob/master/README.md) on running flotilla.
* I created an ECS EC2 cluster and not a Fargate one called flotilla-ec2-east
* I setup my credentials (I also got the same results with using full admin AWS creds too so its not the policy or user account.)
```  
export AWS_ACCESS_KEY_ID=$(aws --profile flotilla configure get aws_access_key_id)
export AWS_SECRET_ACCESS_KEY=$(aws --profile flotilla configure get aws_secret_access_key)
```
* ran `docker-compose up`

I tried two tasks.
* Simple bash `echo hello world` ie Task that terminates
* Bash loop that prints out the date to stderr ie Task that runs forever

## Hello World
The command i used was a one liner
```
echo hello world
```

### Flotilla Task definition
<img width="1407" alt="48507621-47ae0b80-e801-11e8-8fa8-63361fa69372" src="https://user-images.githubusercontent.com/44782696/48640077-70154180-e98a-11e8-8624-7eb990f2cd12.png">

### Flotilla Task Run
<img width="441" alt="48507674-6d3b1500-e801-11e8-9839-865a45167f59" src="https://user-images.githubusercontent.com/44782696/48640148-9d61ef80-e98a-11e8-8908-234db0a9eab3.png">

### ECS Task console
<img width="656" alt="hello_world_ecs_task" src="https://user-images.githubusercontent.com/44782696/48507705-83e16c00-e801-11e8-8e85-b057e512834c.png">

### Cloudwatch logs
<img width="1005" alt="hello_world_cloudwatch_log" src="https://user-images.githubusercontent.com/44782696/48507711-89d74d00-e801-11e8-8606-8abd34f57c22.png">

## Bash Loop
```
cat <<EOF | bash
while true; do
  sleep 1
  date 2>&1
done
EOF
```

### Flotilla Run
<img width="1411" alt="bash_loop_run" src="https://user-images.githubusercontent.com/44782696/48507898-0538fe80-e802-11e8-8355-4576a8d64aea.png">

### Flotilla Task Defintion
<img width="1435" alt="bash_loop_task" src="https://user-images.githubusercontent.com/44782696/48507884-fce0c380-e801-11e8-99cf-156ecb5d83d9.png">

### ECS Task
<img width="1189" alt="bash_loop_ecs_task" src="https://user-images.githubusercontent.com/44782696/48507895-04a06800-e802-11e8-9ade-f3fc929e4e3f.png">

### ECS Task Details
<img width="1154" alt="bash_loop_ecs_task_details" src="https://user-images.githubusercontent.com/44782696/48507896-04a06800-e802-11e8-8906-797ac8eec59c.png">

### ECS Task Logs
<img width="675" alt="bash_loop_ecs_task_logs" src="https://user-images.githubusercontent.com/44782696/48507897-04a06800-e802-11e8-9093-56fe67074676.png">

**Comment by usman-r** [comment]
Hey, just need a bit more info. Could I see the Cloudwatch rule that routes the ECS task status events to the Flotilla status queues, as well as the SQS queues?

**Comment by aaj-st** [comment]
The rule should work because it goes from queued to pending meaning its getting the notifications
![image](https://user-images.githubusercontent.com/44782696/48510102-ba21ea00-e807-11e8-9d4c-b37cf2e805fd.png)

The queue on the other hand, might have something different than expected.
I do notice two queues
![image](https://user-images.githubusercontent.com/44782696/48510311-5a780e80-e808-11e8-9268-7171f78d4b17.png)

![image](https://user-images.githubusercontent.com/44782696/48510432-c195c300-e808-11e8-8e11-7e606f098268.png)


```
{"instance":{"dns_name":"","instance_id":""},"task_arn":"","run_id":"ddce1fe9-9aa1-43ae-7350-403b23e84d0e","definition_id":"FlotillaGroup-69a83433-9939-47fb-663a-a1dc978c2719","alias":"BashLoop","image":"ubuntu:xenial","cluster":"flotilla-ec2-east","status":"QUEUED","group_name":"FlotillaGroup","user":"v1-unknown","env":[{"name":"FLOTILLA_RUN_ID","value":"ddce1fe9-9aa1-43ae-7350-403b23e84d0e"},{"name":"FLOTILLA_RUN_OWNER_ID","value":"v1-unknown"},{"name":"FLOTILLA_SERVER_MODE","value":"dev"}]}
```

This queue is empty
![image](https://user-images.githubusercontent.com/44782696/48510452-ce1a1b80-e808-11e8-8712-980a70b50dce.png)

After i stop polling the flotilla backend does this.
![screen shot 2018-11-14 at 12 34 43 pm](https://user-images.githubusercontent.com/44782696/48510780-b000eb00-e809-11e8-8136-bfd35da7a459.png)


I start polling again on both queues and I get nothing.
My config is the [default config](https://github.com/stitchfix/flotilla-os/blob/master/conf/config.yml)

**Comment by thedatachef** [comment]
@aaj-st two queues is expected. 

The first queue (`dev-flotilla-flotilla-ec2-east`) is created by flotilla to hold runs  _before_ submitting them to ecs. The run worker pulls messages from there and submits them to ecs as tasks. At that point (not via the status queue) does the run go from `QUEUED` to `PENDING` (or `STOPPED`) based on what gets returned from the ecs submit task call. This appears to be working for you.

The second queue is for status updates from ecs about task transitions on the cluster. It looks like the message of transition from `PENDING` to `RUNNING` (and to `STOPPED`) isn't coming? Can we see the `monitoring` tab for the `flotilla-status-updates-dev` queue?

**Comment by thedatachef** [comment]
@aaj-st and the `permissions` tab too please

**Comment by aaj-st** [comment]
@thedatachef 

I had to recreate the ECS cluster again.  I also ran it with my root level permissions to rule out anything to do with policy and permissions at least on starting the local docker jobs ie i didnt use the flotilla policy.

## dev-flotilla-flotilla-ec2-east
### monitoring
<img width="1200" alt="screen shot 2018-11-16 at 9 21 51 am" src="https://user-images.githubusercontent.com/44782696/48637026-60ddc600-e981-11e8-88e9-5d11a99df445.png">
### permissions
<img width="1187" alt="screen shot 2018-11-16 at 9 22 25 am" src="https://user-images.githubusercontent.com/44782696/48637072-79e67700-e981-11e8-927e-f7fda7491577.png">

## flotilla-status-updates-dev
### monitoring
<img width="1159" alt="screen shot 2018-11-16 at 9 22 09 am" src="https://user-images.githubusercontent.com/44782696/48637128-971b4580-e981-11e8-8f91-55b5fab482d4.png">
### permissions
<img width="1185" alt="screen shot 2018-11-16 at 9 22 33 am" src="https://user-images.githubusercontent.com/44782696/48637135-a1d5da80-e981-11e8-812e-3e0ea51a739e.png">

## Flotilla logs
![image](https://user-images.githubusercontent.com/44782696/48637204-d6e22d00-e981-11e8-96e4-34acfe21668c.png)

There doesnt appear to be verbose logging here.  I'd really like to see things like.  sending json request to AWS, waiting on ECS response, got response and then show it.  It does seem like the queue is empty of any updates.

Im also using the default docker-compose.yml and conf/config.yaml

**Comment by thedatachef** [comment]
@aaj-st ok - it looks like the permissions allowing the cloudwatch rule to route status updates into the queue isn't there. If you're blocked, you can add this manually. I'll create an issue for this and work through it 
![sqs-permissions](https://user-images.githubusercontent.com/166823/48638045-583abf00-e984-11e8-9493-92ff2dcb6636.jpg)

Can you please open an issue for the logging if it's not what you expect it should be? Ideally with specific examples and details. Thanks!

**Comment by thedatachef** [comment]
There were a couple underlying issues here - specifically there were a few permissions that weren't listed in the policy document in the readme, most notably (`sqs:SetQueueAttributes`). 

Second, when the service initially started up it created the routing rule correctly but failed with a permissions issue on allowing the status update events to flow to the sqs queue. Since the code path there wasn't correctly idempotent the error message never resurfaced on subsequent restarts. I've added a check for this and verified this works. Closing
